src/ContainerTrait.php

```diff
@@ -2,34 +2,213 @@
 
 namespace atk4\core;
 
+/**
+ * This trait makes it possible for you to add child objects
+ * into your object.
+ */
 trait ContainerTrait {
 
+    /**
+     * Check this property to see if ContainerTrait is present
+     * in the object
+     *
+     * @var string
+     */
+    public $_containerTrait = true;
+
+    /**
+     * Unique object name
+     *
+     * @var string
+     */
+    public $name;
+
+    /**
+     * short_name => object hash of children objects. If the child is not
+     * trackable, then object will be set to "true" (to avoid extra reference)
+     *
+     * @var array
+     */
     protected $elements;
 
     private $_element_name_counts = array();
-    public function _unique_element($desired = null)
+    public function _unique_element($desired)
     {
         $postfix = @++$this->_element_name_counts[$desired];
 
         return $desired.($postfix > 1 ? ('_'.$postfix) : '');
     }
 
-    protected function _unique() {
+    /**
+     * If you are using ContainerTrait only, then you can safely
+     * use this add() method. If you are also using factory, or
+     * initializer then redefine add() and call
+     * _add_Container, _add_Factory, 
+     */
+    function add($obj, $args = [])
+    {
+        if (isset($this->_factoryTrait)) {
+            // Factory allows us to pass string-type objects
+            $obj = $this->_add_Factory($obj, $args);
```

Comment:
Aren't we following some king of CS for methods naming i.e. shouldn't $this->_add_Factory be $this->_addFactory?